source/agora/consensus/protocol/Nominator.d

```diff
@@ -1031,24 +1033,86 @@ extern(D):
+        import agora.common.Amount;
+
+        /// Used for holding consensus candidate values. It contains precomputed
+        /// fields to speed up sorting.
+        static struct CandidateHolder
+        {
+            /// Consensus data
+            public ConsensusData consensus_data;
+            /// Hash of the consensus data
+            public Hash hash;
+            /// The total amount of fees of the transactions in the consensus data
+            public Amount total_adjusted_fee;
+
+            /// Comparison function, which sorts by
+            /// 1. length of missing validators (smallest first), or if it ties
+            /// 2. total adjusted fee (smallest last), or if it ties
+            /// 3. hash of the candidate (smallest first)
+            public int opCmp (in CandidateHolder other) const @safe scope pure nothrow @nogc
+            {
+                if (this.consensus_data.missing_validators.length <
+                    other.consensus_data.missing_validators.length)
+                        return -1;
+                    else if (this.consensus_data.missing_validators.length >
+                             other.consensus_data.missing_validators.length)
+                        return 1;
+
+                    if (this.total_adjusted_fee > other.total_adjusted_fee)
+                        return -1;
+                    else if (this.total_adjusted_fee < other.total_adjusted_fee)
+                        return 1;
+
+                    if (this.hash < other.hash)
+                        return -1;
+                    else if (this.hash > other.hash)
+                        return 1;
+
+                    return 0;
+            }
+        }
+
         try
         {
-            ConsensusData[] values;
+            CandidateHolder[] candidate_holders;
             foreach (ref const(Value) candidate; candidates)
             {
                 auto data = deserializeFull!ConsensusData(candidate[]);
+                log.trace("Consensus data: {}", data.prettify);
 
+                // Only allowed in unittests, as validating the consensus data
+                // for all the candidates might take long.
+                version (unittest)
                 if (auto msg = this.ledger.validateConsensusData(data))
                     assert(0, format!"combineCandidates: Invalid consensus data: %s"(
                         msg));
 
-                log.trace("Combined consensus data: {}", data.prettify);
-                values ~= data;
+                Amount total_adjusted_fee;
+                foreach (const ref tx_hash; data.tx_set)
+                {
+                    Amount adjusted_fee;
+                    auto errormsg = this.ledger.getAdjustedTXFee(tx_hash, adjusted_fee);
+                    if (errormsg == Ledger.InvalidConsensusDataReason.NotInPool)
+                        continue; // most likely a CoinBase Transaction
+                    else if (errormsg)
+                        assert(0);
```

Comment:
```D
if (auto msg = ...)
{
    assert(msg == Ledger.InvalidConsensusDataReason.NotInPool);
    continue;
}
```

Reply:
that is not equivalent with my code though, as I want to assert even in release mode

Reply:
As mentioned on slack, if we ever ship with `-release`, we will need to go over all the asserts anyway.